src/providers/tpm_provider/mod.rs

```diff
@@ -0,0 +1,462 @@
+        Err(ResponseStatus::KeyIDManagerError)
+    })?;
+    let password_context = match password_context {
+        Some(context) => context,
+        None => {
+            error!(
+                "Key triple \"{}\" does not exist in the Key ID Manager.",
+                key_triple
+            );
+            return Err(ResponseStatus::KeyDoesNotExist);
+        }
+    };
+    Ok(bincode::deserialize(password_context).or_else(|e| {
+        error!("Error deserializing the PasswordContext: {}.", e);
+        Err(ResponseStatus::KeyIDManagerError)
+    })?)
+}
+
+impl TpmProvider {
+    /// Creates and initialise a new instance of TpmProvider.
+    fn new(
+        key_id_store: Arc<RwLock<dyn ManageKeyIDs + Send + Sync>>,
+        esapi_context: tss_esapi::TransientObjectContext,
+    ) -> Option<TpmProvider> {
+        Some(TpmProvider {
+            esapi_context: Mutex::new(esapi_context),
+            key_id_store,
+        })
+    }
+}
+
+impl Provide for TpmProvider {
+    fn list_opcodes(&self, _op: OpListOpcodes) -> Result<ResultListOpcodes> {
+        Ok(ResultListOpcodes {
+            opcodes: SUPPORTED_OPCODES.iter().copied().collect(),
+        })
+    }
+
+    fn describe(&self) -> ProviderInfo {
+        ProviderInfo {
+            // Assigned UUID for this provider: 1e4954a4-ff21-46d3-ab0c-661eeb667e1d
+            uuid: Uuid::parse_str("1e4954a4-ff21-46d3-ab0c-661eeb667e1d").expect("UUID parsing failed"),
+            description: String::from("TPM provider, interfacing with a library implementing the TCG TSS 2.0 Enhanced System API specification."),
+            vendor: String::from("Trusted Computing Group (TCG)"),
+            version_maj: 0,
+            version_min: 1,
+            version_rev: 0,
+            id: ProviderID::TpmProvider,
+        }
+    }
+
+    fn create_key(&self, app_name: ApplicationName, op: OpCreateKey) -> Result<ResultCreateKey> {
+        if op.key_attributes.key_type != KeyType::RsaKeypair
+            || op.key_attributes.algorithm != Algorithm::sign(SignAlgorithm::RsaPkcs1v15Sign, None)
+        {
+            error!("The TPM provider currently only supports creating RSA key pairs for signing and verifying.");
+            return Err(ResponseStatus::UnsupportedOperation);
+        }
+
+        let key_name = op.key_name;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+        // This should never panic on 32 bits or more machines.
+        let key_size = std::convert::TryFrom::try_from(op.key_attributes.key_size)
+            .expect("Conversion to usize failed.");
+
+        let mut store_handle = self.key_id_store.write().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let (key_context, auth_value) = esapi_context
+            .create_rsa_signing_key(key_size, AUTH_VAL_LEN)
```

Comment:
I don't know if creating an "RSA key definition structure" and passing that here is more sensible. We should be setting the hash algorithm, for example.

Maybe we could have a method like `create_signing_key` which takes in an enum (`RSA` or `ECC`) and the enum contains the other params as well... Don't know, really

Reply:
I think it is fine for now as we are limiting the scope of things that we support.

Reply:
True. We'll have to see how to manage that for the TPM crate, though